src/set_theory/fixed_points.lean

```diff
@@ -0,0 +1,508 @@
+
+## Main definitions and results
+
+* `nfp_family`, `nfp_bfamily`, `nfp`: the next fixed point of a (family of) normal function(s).
+* `fp_family_unbounded`, `fp_bfamily_unbounded`, `fp_unbounded`: the (common) fixed points of a
+  (family of) normal function(s) are unbounded in the ordinals.
+* `deriv_add_eq_mul_omega_add`: a characterization of the derivative of addition.
+-/
+
+noncomputable theory
+
+universes u v
+
+open function
+
+namespace ordinal
+
+/-! ### Fixed points of type-indexed families of ordinals -/
+
+section
+variables {ι : Type u} {f : ι → ordinal.{max u v} → ordinal.{max u v}}
+
+/-- Applies the functions specified by the indices of a list, in order, to a specified value. -/
+def nfp_family_iterate (f : ι → ordinal → ordinal) (a : ordinal) : list ι → ordinal
+| []       := a
+| (i :: l) := f i (nfp_family_iterate l)
+
+@[simp] theorem nfp_family_iterate_nil (f : ι → ordinal → ordinal) (a) :
+  nfp_family_iterate f a [] = a :=
+rfl
+
+@[simp] theorem nfp_family_iterate_cons (f : ι → ordinal → ordinal) (i l a) :
+  nfp_family_iterate f a (i :: l) = f i (nfp_family_iterate f a l) :=
+rfl
+
+theorem nfp_family_iterate_empty [is_empty ι] (f : ι → ordinal → ordinal) (a) :
+  Π l : list ι, nfp_family_iterate f a l = a
+| []       := rfl
+| (i :: l) := is_empty_elim i
+
+theorem nfp_family_iterate_fixed {a} (ha : ∀ i, f i a = a) :
+  Π l : list ι, nfp_family_iterate f a l = a
+| []       := rfl
+| (i :: l) := by { convert ha i, exact congr_arg (f i) (nfp_family_iterate_fixed l) }
+
+theorem nfp_family_iterate_eq_iterate (f : ordinal → ordinal) (a) :
+  Π l, nfp_family_iterate (λ _ : ι, f) a l = (f^[l.length]) a
+| []       := rfl
+| (i :: l) := begin
+  convert congr_arg f (nfp_family_iterate_eq_iterate l),
+  exact function.iterate_succ_apply' f _ a
+end
+
+theorem nfp_family_iterate_monotone {f : ι → ordinal → ordinal} (H : ∀ i, monotone (f i))
+  (l : list ι) : monotone (λ a, nfp_family_iterate f a l) :=
+λ a b h, list.rec_on l h (λ i l hl, H i hl)
+
+/-- The next common fixed point, at least `a`, for a family of normal functions. -/
+def nfp_family (f : ι → ordinal → ordinal) (a) : ordinal :=
+sup (nfp_family_iterate f a)
+
+theorem nfp_family_eq_sup (f : ι → ordinal → ordinal) (a) :
+  nfp_family f a = sup (nfp_family_iterate f a) :=
+rfl
+
+theorem iterate_le_nfp_family (f : ι → ordinal → ordinal) (a l) :
+  nfp_family_iterate f a l ≤ nfp_family f a :=
+le_sup _ _
+
+theorem self_le_nfp_family (f : ι → ordinal → ordinal) (a) : a ≤ nfp_family f a :=
+le_sup _ []
+
+theorem lt_nfp_family {a b} : a < nfp_family f b ↔ ∃ l, a < nfp_family_iterate f b l :=
+lt_sup
+
+theorem nfp_family_le {a b} : nfp_family f a ≤ b ↔ ∀ l, nfp_family_iterate f a l ≤ b :=
+sup_le
+
+theorem nfp_family_monotone (hf : ∀ i, monotone (f i)) : monotone (nfp_family f) :=
+λ a b h, sup_le.2 $ λ l, (nfp_family_iterate_monotone hf l h).trans (le_sup _ l)
+
+theorem apply_lt_nfp_family (H : ∀ i, is_normal (f i)) {a b} (hb : b < nfp_family f a) (i) :
```

Comment:
I wonder if there is a better name for this lemma. Also, can this be proved more directly by arguing that b <= f i b, because `f i` is normal?

Reply:
What you mention is what `apply_lt_nfp_family_iff` does. As for the name, I have nothing better. Is there a "standard" mathlib name for `f i` to use instead of `apply`?